test/integration/login_test.go

```diff
@@ -0,0 +1,152 @@
+	err = loginOptions.GatherServerInfo()
+	if err == nil {
+		t.Fatalf("Raw login should error out")
+	} else {
+		if !strings.Contains(err.Error(), "OpenShift is not configured") {
+			t.Fatalf("Raw login should display OpenShift not configured message")
+		}
+	}
+
+	username := "joe"
+	password := "pass"
+	project := "the-singularity-is-near"
+	server := openshiftClientConfig.Host
+
+	loginOptions = newLoginOptions(server, username, password, "", true)
+
+	if err = loginOptions.GatherServerInfo(); err != nil {
+		t.Fatalf("Error trying to determine server info: ", err)
+	}
+
+	if err = loginOptions.GatherAuthInfo(); err != nil {
+		t.Fatalf("Error trying to determine auth info: ", err)
+	}
+
+	me, err := whoami(loginOptions.Config)
+	if err != nil {
+		t.Fatalf("unexpected error: ", err)
+	}
+	if me.Name != "anypassword:"+username {
+		t.Fatalf("Unexpected user after authentication: %v", me.Name)
+	}
+
+	newProjectOptions := &newproject.NewProjectOptions{
+		Client:                openshiftClient,
+		ProjectName:           project,
+		AdminRole:             "admin",
+		MasterPolicyNamespace: "master",
+		AdminUser:             "anypassword:" + username,
+	}
+	if err := newProjectOptions.Run(); err != nil {
+		t.Fatalf("unexpected error: ", err)
+	}
+
+	oClient, _ := client.New(loginOptions.Config)
+	p, err := oClient.Projects().Get(project)
+	if err != nil {
+		t.Fatalf("unexpected error: ", err)
+	}
+
+	if p.Name != project {
+		t.Fatalf("Got the unexpected project: %v", p.Name)
+	}
+
+	// TODO Commented because of incorrectly hitting cache when listing projects.
+	// Should be enabled again when cache eviction is properly fixed.
+
+	// err = loginOptions.GatherProjectInfo()
+	// if err != nil {
+	// 	t.Fatalf("unexpected error: ", err)
+	// }
+
+	// if loginOptions.Project != project {
+	// 	t.Fatalf("Expected project %v but got %v", project, loginOptions.Project)
+	// }
+
+	// configFile, err := ioutil.TempFile("", "openshiftconfig")
+	// if err != nil {
+	// 	t.Fatalf("unexpected error: %v", err)
+	// }
+	// defer os.Remove(configFile.Name())
+
+	// if _, err = loginOptions.SaveConfig(configFile.Name()); err != nil {
+	// 	t.Fatalf("unexpected error: ", err)
+	// }
+}
+
+func newLoginOptions(server string, username string, password string, context string, insecure bool) *cmd.LoginOptions {
+	flagset := pflag.NewFlagSet("test-flags", pflag.ContinueOnError)
+	factory := clientcmd.New(flagset)
+
+	flags := []string{}
+
+	if len(server) > 0 {
+		flags = append(flags, "--server="+server)
+	}
+	if len(context) > 0 {
+		flags = append(flags, "--context="+context)
+	}
+	if insecure {
+		flags = append(flags, "--insecure-skip-tls-verify")
+	}
+
+	flagset.Parse(flags)
+
+	loginOptions := &cmd.LoginOptions{
+		ClientConfig: factory.OpenShiftClientConfig,
+		Reader:       os.Stdin,
+		Username:     username,
+		Password:     password,
+	}
+
+	return loginOptions
+}
+
+func whoami(clientCfg *kclient.Config) (*api.User, error) {
```

Comment:
Exposed and actually became a method of `LoginOptions` that uses its internal pointer to `Config`.